nixos/modules/services/networking/netbird.nix

```diff
@@ -1,171 +1,504 @@
+              };
+
+              ui.enable = mkOption {
+                type = bool;
+                default = nixosConfig.services.netbird.ui.enable;
+                defaultText = literalExpression ''config.ui.enable'';
+                description = ''
+                  Controls presence of `netbird-ui` wrapper for this Netbird client.
+                '';
+              };
+
+              wrapper = mkOption {
+                type = package;
+                internal = true;
+                default =
+                  let
+                    makeWrapperArgs = concatLists (mapAttrsToList
+                      (key: value: [ "--set-default" key value ])
+                      config.environment
+                    );
+                  in
+                  pkgs.stdenv.mkDerivation {
+                    name = "${cfg.package.name}-wrapper-${name}";
+                    meta.mainProgram = "netbird-${name}";
+                    nativeBuildInputs = with pkgs; [ makeWrapper ];
+                    phases = [ "installPhase" ];
+                    installPhase = concatStringsSep "\n" [
+                      ''
+                        mkdir -p "$out/bin"
+                        makeWrapper ${lib.getExe cfg.package} "$out/bin/netbird-${name}" \
+                          ${escapeShellArgs makeWrapperArgs}
+                      ''
+                      (optionalString cfg.ui.enable ''
+                        # netbird-ui doesn't support envvars
+                        makeWrapper ${lib.getExe cfg.ui.package} "$out/bin/netbird-ui-${name}" \
+                          --add-flags '--daemon-addr=${config.environment.NB_DAEMON_ADDR}'
+
+                        mkdir -p "$out/share/applications"
+                        substitute ${cfg.ui.package}/share/applications/netbird.desktop \
+                            "$out/share/applications/netbird-${name}.desktop" \
+                          --replace-fail 'Name=Netbird' "Name=Netbird @ ${config.name}" \
+                          --replace-fail '${lib.getExe cfg.ui.package}' "$out/bin/netbird-ui-${name}"
+                      '')
+                    ];
+                  };
+              };
+
+              # see https://github.com/netbirdio/netbird/blob/88747e3e0191abc64f1e8c7ecc65e5e50a1527fd/client/internal/config.go#L49-L82
+              config = mkOption {
+                type = attrsOf anything;
+                defaultText = literalExpression ''
+                  {
+                    DisableAutoConnect = !config.autoStart;
+                    WgIface = config.interface;
+                    WgPort = config.port;
+                  }
+                '';
+                description = ''
+                  Additional configuration that exists before the first start and
+                  later overrides the existing values in `config.json`.
+
+                  WARNING: this is not an upstream feature, see below for details.
```

Comment:
I'd emphasize that it can break at any moment

Reply:
I would see an upstream feature built on top of [`ConfigInput` structure ](https://github.com/netbirdio/netbird/pull/1586/files#diff-a669c9314247b993250fdcde6c802c8f689df6e5ce935ad890a4548c7b380a9aR39) currently used for modifying configs by CLI flags.

It would use a different format and possibly have higher priority due to being executed later in the startup process.

This NixOS option would still be useful for configurations not supported by `ConfigInput`.

Reply:
The only valid way of configuring netbird is via environment variables or cli options.The `config.json` is not a config, but a state file and it's expected to be modified only by netbird itself. If we expose a footgun it should be marked as such.